promql/parser/parse_test.go

```diff
@@ -494,12 +494,12 @@ var testExpr = []struct {
 	{
 		input:  "2.5.",
 		fail:   true,
-		errMsg: "unexpected character: '.'",
+		errMsg: `1:1: parse error: bad number or duration syntax: "2.5"`,
```

Comment:
Before this patch, unexpected suffixes including dots, exponents, or underscores returned [`true`](https://github.com/rexagod/prometheus/blob/92f463a51d188823ca37ab762e135e70e837f8f6/promql/parser/lex.go#L918), eventually logging the [previous error](https://github.com/openshift/prometheus/blob/b49a80354a59ff1010a9f1708c3ac0ef27632345/promql/parser/lex.go#L494).

With this change, these three patterns are **completely** handled by the function itself, which now returns `false` under such a case, instead of deferring to [`l.next`'s evaluation](https://github.com/openshift/prometheus/blob/b49a80354a59ff1010a9f1708c3ac0ef27632345/promql/parser/lex.go#L384), since we know beforehand when there is an error.

That being said, I would personally prefer `2.5.` instead of `2.5` in the error message.

Reply:
(fixed)